ClangSharp.PInvokeGenerator/PInvokeGenerator.cs

```diff
@@ -3,43 +3,119 @@
 
         private OutputBuilder _outputBuilder;
         private int _outputBuilderUsers;
+        private bool _disposed;
 
-        public CursorWriter(ConfigurationOptions config)
+        public PInvokeGenerator(PInvokeGeneratorConfiguration config, Func<string, (Stream stream, bool leaveOpen)> outputStreamFactory = null)
         {
-            _outputBuilders = new Dictionary<string, OutputBuilder>();
+            if (config is null)
+            {
+                throw new ArgumentNullException(nameof(config));
+            }
+
+            _index = CXIndex.Create();
+            _outputBuilderFactory = new OutputBuilderFactory();
+            _outputStreamFactory = outputStreamFactory ?? ((path) => {
+                var directoryPath = Path.GetDirectoryName(path);
+                Directory.CreateDirectory(directoryPath);
+                return (new FileStream(path, FileMode.Create), leaveOpen: false);
+            });
             _visitedCursors = new HashSet<Cursor>();
+            _diagnostics = new List<Diagnostic>();
             _config = config;
         }
 
-        public void Dispose()
+        ~PInvokeGenerator()
         {
-            Debug.Assert(_outputBuilder is null);
+            Dispose(isDisposing: false);
+        }
+
+        public PInvokeGeneratorConfiguration Config => _config;
+
+        public IReadOnlyList<Diagnostic> Diagnostics => _diagnostics;
+
+        public CXIndex IndexHandle => _index;
 
-            foreach (var outputBuilder in _outputBuilders)
+        public void Close()
+        {
+            foreach (var outputBuilder in _outputBuilderFactory.OutputBuilders)
             {
-                Debug.Assert(outputBuilder.Key.Equals(outputBuilder.Value.OutputFile));
-                outputBuilder.Value.Dispose();
+                var outputPath = _config.OutputLocation;
+                var isMethodClass = _config.MethodClassName.Equals(outputBuilder.Name);
+
+                if (_config.GenerateMultipleFiles)
+                {
+                    outputPath = Path.Combine(outputPath, $"{outputBuilder.Name}.cs");
+                }
+                else if (isMethodClass)
+                {
+                    outputPath = Path.ChangeExtension(outputPath, $"{_config.MethodClassName}{Path.GetExtension(outputPath)}");
+                }
+
+                var (stream, leaveStreamOpen) = _outputStreamFactory(outputPath);
+                CloseOutputBuilder(stream, outputBuilder, isMethodClass, leaveStreamOpen);
             }
 
-            _outputBuilders.Clear();
+            _diagnostics.Clear();
+            _outputBuilderFactory.Clear();
             _visitedCursors.Clear();
         }
 
-        public void Visit(TranslationUnit translationUnit)
+        public void Dispose()
+        {
+            Dispose(isDisposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        public void GenerateBindings(CXTranslationUnit translationUnitHandle)
```

Comment:
This is now the primary entry point for the library and is what the executable calls into.